File: trustdevice/src/main/java/cn/tongdun/mobrisk/core/FMCore.java
package cn.tongdun.mobrisk.core;

import android.annotation.SuppressLint;
import android.app.ActivityManager;
import android.content.Context;
import android.hardware.SensorManager;

import org.json.JSONObject;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import cn.tongdun.mobrisk.TDRiskCallback;
import cn.tongdun.mobrisk.TDRiskOption;
import cn.tongdun.mobrisk.beans.DeviceInfo;
import cn.tongdun.mobrisk.core.collectors.AppListCollector;
import cn.tongdun.mobrisk.core.collectors.BatteryInfoCollector;
import cn.tongdun.mobrisk.core.collectors.BuildInfoCollector;
import cn.tongdun.mobrisk.core.collectors.CpuInfoCollector;
import cn.tongdun.mobrisk.core.collectors.DebugInfoCollector;
import cn.tongdun.mobrisk.core.collectors.DeviceBaseInfoCollector;
import cn.tongdun.mobrisk.core.collectors.DeviceIdCollector;
import cn.tongdun.mobrisk.core.collectors.DevicePersonalizationInfoCollector;
import cn.tongdun.mobrisk.core.collectors.MagiskCollector;
import cn.tongdun.mobrisk.core.collectors.MemoryInfoCollector;
import cn.tongdun.mobrisk.core.collectors.RootCollector;
import cn.tongdun.mobrisk.core.collectors.SensorsInfoCollector;
import cn.tongdun.mobrisk.core.collectors.SettingInfoCollector;
import cn.tongdun.mobrisk.core.collectors.XposedCollector;
import cn.tongdun.mobrisk.core.utils.DeviceInfoUtils;
import cn.tongdun.mobrisk.core.utils.LogUtils;


public class FMCore {

    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final CountDownLatch countDownLatch = new CountDownLatch(1);
    private Context mContext;
    private DeviceInfo mDeviceInfo;
    private String mDeviceId;
    private TDRiskCallback mCallback = null;

    private FMCore() {
    }

    private static class SingletonHolder {
        @SuppressLint("StaticFieldLeak")
        private static final FMCore instance = new FMCore();
    }

    public static FMCore getInstance() {
        return SingletonHolder.instance;
    }

    public void init(Context context, TDRiskOption option) {
        if (context == null) {
            return;
        }
        if (mDeviceInfo == null) {
            mDeviceInfo = new DeviceInfo();
        }
        if (option != null) {
            mCallback = option.getCallback();
        }
        mContext = context.getApplicationContext();
        executor.execute(() -> {
            collectorDeviceId();
            collectorDebugInfo();
            collectorRoot();
            collectorXposed();
            collectorMagisk();
            collectorBuildInfo();
            collectorDeviceBaseInfo();
            collectorDevicePersonalizationInfo();
            collectorMemoryInfo();
            collectorCpuInfo();
            collectorBatteryInfo();
            collectorSettingInfo();
            collectorSensorInfo();
            collectorPackageListInfo();
            countDownLatch.countDown();
            if (mCallback !=null){
                mCallback.onEvent(getDeviceInfo());
            }
        });
    }

    private void collectorDeviceId() {
        DeviceIdCollector deviceIdCollector = new DeviceIdCollector(mContext.getContentResolver());
        mDeviceId = deviceIdCollector.getDeviceId();
        mDeviceInfo.setAndroidId(deviceIdCollector.getAndroidId());
        mDeviceInfo.setGsfId(deviceIdCollector.getGsfId());
        mDeviceInfo.setMediaDrmId(deviceIdCollector.getMediaDrmId());
        mDeviceInfo.setVbMetaDigest(deviceIdCollector.getVbMetaDigest());
    }

    private void collectorDebugInfo() {
        DebugInfoCollector debugInfoCollector = new DebugInfoCollector();
        mDeviceInfo.setDebug(debugInfoCollector.getDebug());
    }

    private void collectorRoot() {
        RootCollector rootCollector = new RootCollector();
        mDeviceInfo.setRoot(rootCollector.getRoot());
    }

    private void collectorXposed() {
        XposedCollector xposedCollector = new XposedCollector();
        mDeviceInfo.setXposedStatus(xposedCollector.findXposed());
    }

    private void collectorMagisk() {
        MagiskCollector magiskCollector = new MagiskCollector();
        mDeviceInfo.setMagiskStatus(magiskCollector.detectMagisk());
    }

    private void collectorBuildInfo() {
        BuildInfoCollector buildInfoCollector = new BuildInfoCollector();
        mDeviceInfo.setModel(buildInfoCollector.getModel());
        mDeviceInfo.setManufacturer(buildInfoCollector.getManufacturer());
        mDeviceInfo.setAndroidVersion(buildInfoCollector.getAndroidVersion());
        mDeviceInfo.setSdkVersion(buildInfoCollector.getSdkVersion());
        mDeviceInfo.setKernelVersion(buildInfoCollector.getKernelVersion());
        mDeviceInfo.setFingerprint(buildInfoCollector.getFingerprint());
        mDeviceInfo.setHardware(buildInfoCollector.getHardware());
        mDeviceInfo.setProduct(buildInfoCollector.getProduct());
        mDeviceInfo.setBrand(buildInfoCollector.getBrand());
        mDeviceInfo.setHost(buildInfoCollector.getHost());
        mDeviceInfo.setDisplay(buildInfoCollector.getDisplay());
        mDeviceInfo.setHost(buildInfoCollector.getHost());
    }

    private void collectorDeviceBaseInfo() {
        DeviceBaseInfoCollector deviceBaseInfoCollector = new DeviceBaseInfoCollector(mContext);
        mDeviceInfo.setScreenResolution(deviceBaseInfoCollector.getScreenResolution());
        mDeviceInfo.setScreenInches(deviceBaseInfoCollector.getScreenInches());
        mDeviceInfo.setFilesAbsolutePath(deviceBaseInfoCollector.getFilesAbsolutePath());
        mDeviceInfo.setPackageName(deviceBaseInfoCollector.getPackageName());
        mDeviceInfo.setHarmonyOS(deviceBaseInfoCollector.isHarmonyOS());
    }

    private void collectorDevicePersonalizationInfo() {
        DevicePersonalizationInfoCollector devicePersonalizationInfoCollector = new DevicePersonalizationInfoCollector();
        mDeviceInfo.setCountry(devicePersonalizationInfoCollector.getLocaleCountry());
        mDeviceInfo.setLanguage(devicePersonalizationInfoCollector.getDefaultLanguage());
        mDeviceInfo.setTimezone(devicePersonalizationInfoCollector.getTimezone());
    }

    private void collectorMemoryInfo() {
        ActivityManager activityManager = (ActivityManager) mContext.getSystemService(Context.ACTIVITY_SERVICE);
        MemoryInfoCollector memoryInfoCollector = new MemoryInfoCollector(activityManager);
        mDeviceInfo.setTotalMemory(String.valueOf(memoryInfoCollector.getTotalMemory()));
        mDeviceInfo.setAvailableMemory(String.valueOf(memoryInfoCollector.getAvailableMemory()));
        mDeviceInfo.setTotalStorage(String.valueOf(memoryInfoCollector.getTotalStorage()));
        mDeviceInfo.setAvailableStorage(String.valueOf(memoryInfoCollector.getAvailableStorage()));
    }

    private void collectorCpuInfo() {
        CpuInfoCollector cpuInfoCollector = new CpuInfoCollector();
        mDeviceInfo.setCoresCount(cpuInfoCollector.availableProcessors());
        mDeviceInfo.setCpuHardware(cpuInfoCollector.getCpuHardware());
        mDeviceInfo.setCpuProcessor(cpuInfoCollector.getCpuProcessor());
        mDeviceInfo.setAbiType(cpuInfoCollector.getAbiType());
    }

    private void collectorBatteryInfo() {
        BatteryInfoCollector batteryInfoCollector = new BatteryInfoCollector(mContext);
        mDeviceInfo.setBatteryHealthStatus(batteryInfoCollector.getBatteryHealthStatus());
        mDeviceInfo.setBatteryStatus(batteryInfoCollector.getBatteryStatus());
        mDeviceInfo.setBatteryLevel(String.valueOf(batteryInfoCollector.getBatteryLevel()));
        mDeviceInfo.setBatteryTemp(String.valueOf(batteryInfoCollector.getBatteryTemp()));
        mDeviceInfo.setBatteryTotalCapacity(batteryInfoCollector.getBatteryTotalCapacity());
    }

    private void collectorSettingInfo() {
        SettingInfoCollector settingInfoCollector = new SettingInfoCollector(mContext.getContentResolver());
        mDeviceInfo.setAdbEnabled(settingInfoCollector.getAdbEnabled());
        mDeviceInfo.setDevelopmentSettingEnabled(settingInfoCollector.getDevelopmentSettingEnabled());
        mDeviceInfo.setHttpProxy(settingInfoCollector.getHttpProxy());
        mDeviceInfo.setDataRoaming(settingInfoCollector.getDataRoaming());
        mDeviceInfo.setAllowMockLocation(settingInfoCollector.getAllowMockLocation());
        mDeviceInfo.setAccessibilityEnabled(settingInfoCollector.getAccessibilityEnabled());
        mDeviceInfo.setDefaultInputMethod(settingInfoCollector.getDefaultInputMethod());
        mDeviceInfo.setTouchExplorationEnabled(settingInfoCollector.getTouchExplorationEnabled());
        mDeviceInfo.setScreenBrightness(settingInfoCollector.getScreenBrightness());
        mDeviceInfo.setScreenOffTimeout(settingInfoCollector.getScreenOffTimeout());
    }

    private void collectorSensorInfo() {
        SensorManager sensorManager = (SensorManager) mContext.getSystemService(Context.SENSOR_SERVICE);
        SensorsInfoCollector sensorsInfoCollector = new SensorsInfoCollector(sensorManager);
        mDeviceInfo.setSensorsInfo(sensorsInfoCollector.getSensorList());
    }

    private void collectorPackageListInfo() {
        AppListCollector appListCollector = new AppListCollector(mContext.getPackageManager());
        mDeviceInfo.setAppList(appListCollector.getAppList());
        mDeviceInfo.setSystemAppList(appListCollector.getSystemAppList());
    }

    public JSONObject getDeviceInfo() {
        try {
            countDownLatch.await(1000L, TimeUnit.MILLISECONDS);
        } catch (Throwable e) {
            LogUtils.e("getDeviceInfo await exception e:" + e.getMessage());
        }
        return DeviceInfoUtils.format(mDeviceId, mDeviceInfo);
    }
}

File: trustdevice/src/main/java/cn/tongdun/mobrisk/core/FMCore.kt
package cn.tongdun.mobrisk.core

import android.app.ActivityManager
import android.content.Context
import android.hardware.SensorManager
import cn.tongdun.mobrisk.TDRiskCallback
import cn.tongdun.mobrisk.TDRiskOption
import cn.tongdun.mobrisk.beans.DeviceInfo
import cn.tongdun.mobrisk.core.collectors.*
import cn.tongdun.mobrisk.core.tools.DeviceInfoUtils
import cn.tongdun.mobrisk.core.collectors.SettingInfoCollector
import cn.tongdun.mobrisk.core.tools.executeSafe
import org.json.JSONObject
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

class FMCore private constructor() {

    companion object {
        init {
            executeSafe { System.loadLibrary("tddevicefingerprint") }
        }

        val instance: FMCore by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) {
            FMCore()
        }
    }


    private val executor = Executors.newSingleThreadExecutor()
    private val countDownLatch = CountDownLatch(1)
    private lateinit var mContext: Context
    private lateinit var mDeviceInfo: DeviceInfo
    private lateinit var mDeviceId: String
    private var mCallback: TDRiskCallback? = null

    fun init(context: Context, option: TDRiskOption) {
        mContext = context.applicationContext
        mDeviceInfo = DeviceInfo()
        mCallback = option.callback
        executor.execute {
            collectorDeviceId()
            collectorDebugInfo()
            collectorRoot()
            collectorXposed()
            collectorMagisk()
            collectorEmulator()
            collectorHook()
            collectorVpn()
            collectorDeviceInfoTampered()
            collectorBuildInfo()
            collectorDeviceBaseInfo()
            collectorDevicePersonalizationInfo()
            collectorMemoryInfo()
            collectorCpuInfo()
            collectorBatteryInfo()
            collectorSettingInfo()
            collectorSensorInfo()
            collectorPackageListInfo()
            countDownLatch.countDown()
            mCallback?.onEvent(getDeviceInfo())
        }
    }

    private fun collectorDeviceId() {
        val deviceIdCollector = DeviceIdCollector(mContext.contentResolver)
        mDeviceId = deviceIdCollector.getDeviceId()
        mDeviceInfo.androidId = deviceIdCollector.getAndroidId()
        mDeviceInfo.gsfId = deviceIdCollector.getGsfId()
        mDeviceInfo.mediaDrmId = deviceIdCollector.getMediaDrmId()
        mDeviceInfo.vbMetaDigest = deviceIdCollector.getVbMetaDigest()
        mDeviceInfo.googleAdId = deviceIdCollector.getGoogleAdid(mContext)
    }

    private fun collectorDebugInfo() {
        val debugInfoCollector = DebugInfoCollector()
        mDeviceInfo.debug = debugInfoCollector.getDebug()
    }

    private fun collectorRoot() {
        val rootCollector = RootCollector()
        mDeviceInfo.root = rootCollector.getRoot()
    }

    private fun collectorXposed() {
        val xposedCollector = XposedCollector()
        mDeviceInfo.xposed = xposedCollector.findXposed()
    }

    private fun collectorMagisk() {
        val magiskCollector = MagiskCollector()
        mDeviceInfo.magisk = magiskCollector.detectMagisk()
    }

    private fun collectorHook() {
        val hookCollector = HookCollector()
        mDeviceInfo.hook = hookCollector.detectHook()
    }

    private fun collectorEmulator() {
        val emulatorCollector = EmulatorCollector()
        mDeviceInfo.emulator = emulatorCollector.detectEmulator()
    }

    private fun collectorVpn() {
        val vpnCollector = VpnCollector()
        mDeviceInfo.vpn = vpnCollector.detectVpn(mContext)
    }

    private fun collectorDeviceInfoTampered() {
        val deviceIdCollector = DeviceInfoTamperedCollector()
        mDeviceInfo.deviceInfoTampered = deviceIdCollector.tampered(mContext)
    }

    private fun collectorBuildInfo() {
        val buildInfoCollector = BuildInfoCollector()
        mDeviceInfo.model = buildInfoCollector.getModel()
        mDeviceInfo.manufacturer = buildInfoCollector.getManufacturer()
        mDeviceInfo.androidVersion = buildInfoCollector.getAndroidVersion()
        mDeviceInfo.sdkVersion = buildInfoCollector.getSdkVersion()
        mDeviceInfo.kernelVersion = buildInfoCollector.getKernelVersion()
        mDeviceInfo.fingerprint = buildInfoCollector.getFingerprint()
        mDeviceInfo.hardware = buildInfoCollector.getHardware()
        mDeviceInfo.product = buildInfoCollector.getProduct()
        mDeviceInfo.brand = buildInfoCollector.getBrand()
        mDeviceInfo.host = buildInfoCollector.getHost()
        mDeviceInfo.display = buildInfoCollector.getDisplay()
    }

    private fun collectorDeviceBaseInfo() {
        val deviceBaseInfoCollector = DeviceBaseInfoCollector(mContext)
        mDeviceInfo.screenResolution = deviceBaseInfoCollector.getScreenResolution()
        mDeviceInfo.screenInches = deviceBaseInfoCollector.getScreenInches()
        mDeviceInfo.filesAbsolutePath = deviceBaseInfoCollector.getFilesAbsolutePath()
        mDeviceInfo.packageName = deviceBaseInfoCollector.getPackageName()
        mDeviceInfo.harmonyOS = deviceBaseInfoCollector.isHarmonyOS()
    }

    private fun collectorDevicePersonalizationInfo() {
        val devicePersonalizationInfoCollector = DevicePersonalizationInfoCollector()
        mDeviceInfo.country = devicePersonalizationInfoCollector.getLocaleCountry()
        mDeviceInfo.language = devicePersonalizationInfoCollector.getDefaultLanguage()
        mDeviceInfo.timezone = devicePersonalizationInfoCollector.getTimezone()
    }

    private fun collectorMemoryInfo() {
        val activityManager = mContext.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val memoryInfoCollector = MemoryInfoCollector(activityManager)
        mDeviceInfo.totalMemory = memoryInfoCollector.getTotalMemory().toString()
        mDeviceInfo.availableMemory = memoryInfoCollector.getAvailableMemory().toString()
        mDeviceInfo.totalStorage = memoryInfoCollector.getTotalStorage().toString()
        mDeviceInfo.availableStorage = memoryInfoCollector.getAvailableStorage().toString()
    }

    private fun collectorCpuInfo() {
        val cpuInfoCollector = CpuInfoCollector()
        mDeviceInfo.coresCount = cpuInfoCollector.availableProcessors()
        mDeviceInfo.cpuHardware = cpuInfoCollector.getCpuHardware()
        mDeviceInfo.cpuProcessor = cpuInfoCollector.getCpuProcessor()
        mDeviceInfo.abiType = cpuInfoCollector.getAbiType()
    }

    private fun collectorBatteryInfo() {
        val batteryInfoCollector = BatteryInfoCollector(mContext)
        mDeviceInfo.batteryHealthStatus = batteryInfoCollector.getBatteryHealth()
        mDeviceInfo.batteryStatus = batteryInfoCollector.getBatteryStatus()
        mDeviceInfo.batteryLevel = batteryInfoCollector.getBatteryLevel().toString()
        mDeviceInfo.batteryTemp = batteryInfoCollector.getBatteryTemp().toString()
        mDeviceInfo.batteryTotalCapacity = batteryInfoCollector.getBatteryTotalCapacity()
    }

    private fun collectorSettingInfo() {
        val settingInfoCollector = SettingInfoCollector(mContext.contentResolver)
        mDeviceInfo.adbEnabled = settingInfoCollector.getAdbEnabled()
        mDeviceInfo.developmentSettingEnabled = settingInfoCollector.getDevelopmentSettingEnabled()
        mDeviceInfo.httpProxy = settingInfoCollector.getHttpProxy()
        mDeviceInfo.dataRoaming = settingInfoCollector.getDataRoaming()
        mDeviceInfo.allowMockLocation = settingInfoCollector.getAllowMockLocation()
        mDeviceInfo.accessibilityEnabled = settingInfoCollector.getAccessibilityEnabled()
        mDeviceInfo.defaultInputMethod = settingInfoCollector.getDefaultInputMethod()
        mDeviceInfo.touchExplorationEnabled = settingInfoCollector.getTouchExplorationEnabled()
        mDeviceInfo.screenBrightness = settingInfoCollector.getScreenBrightness()
        mDeviceInfo.screenOffTimeout = settingInfoCollector.getScreenOffTimeout()
    }

    private fun collectorSensorInfo() {
        val sensorManager = mContext.getSystemService(Context.SENSOR_SERVICE) as SensorManager
        val sensorsInfoCollector = SensorsInfoCollector(sensorManager)
        mDeviceInfo.sensorsInfo = sensorsInfoCollector.getSensorList()
    }

    private fun collectorPackageListInfo() {
        val appListCollector = AppListCollector(mContext.packageManager)
        mDeviceInfo.appList = appListCollector.getAppList()
        mDeviceInfo.systemAppList = appListCollector.getSystemAppList()
    }

    fun getDeviceInfo(): JSONObject {
        executeSafe { countDownLatch.await(1000L, TimeUnit.MILLISECONDS) }
        return DeviceInfoUtils.format(mDeviceId, mDeviceInfo)
    }
}